File: cmd/team/internal/process/process.go
// Package process manages kiro-cli as a subprocess with stdin/stdout pipes.
package process

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"sync"
	"time"
)

const (
	maxCaptureLines = 5000
	responseTimeout = 15 * time.Minute
	timeoutMessage  = "⏱️ 搜尋超時（超過 15 分鐘），請稍後再試或換個問法。"
)

var (
	// ANSI escape sequences
	ansiRegex    = regexp.MustCompile(`\x1b\[[0-9;]*[a-zA-Z]`)
	ansiQRegex   = regexp.MustCompile(`\x1b\[\?[0-9;]*[a-zA-Z]`)
	spinnerRegex = regexp.MustCompile(`[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]\s*\w+\.\.\.`)
	brailleRegex = regexp.MustCompile(`^[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]+\s*`)
	multiSpace   = regexp.MustCompile(`  +`)
)

// skipPatterns are lines to filter from kiro-cli output when extracting responses.
var skipPatterns = []string{
	"▸ Time:", "╭─", "╰─", "│", "Model:",
	"Did you know?", "/context", "/model",
	"All tools are now trusted",
	"ask a question",
	"Thinking...", "Working...",
	"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏",
	"[?25", "[?12",
	"Running tool", "using tool:",
	"Completed in", "- Completed",
	"Successfully read", "Reading file:",
	"Reading directory:", "Querying available",
	"I will run the following command:",
	"Purpose:", "max depth:", "max entries:",
	"(from mcp server:",
	"⋮", "with the param",
}

// ManagedProcess wraps a kiro-cli subprocess with stdin/stdout access.
type ManagedProcess struct {
	Name string

	cmd    *exec.Cmd
	stdin  io.WriteCloser
	stdout io.ReadCloser

	mu           sync.Mutex
	outputLines  []string
	pendingResp  []string
	collecting   bool
	responseCh   chan string
}

// Start launches the kiro-cli subprocess.
func (mp *ManagedProcess) Start(command string, cwd string) error {
	// Parse command into executable and arguments
	parts := parseCommand(command)
	if len(parts) == 0 {
		return fmt.Errorf("空命令")
	}

	cmd := exec.Command(parts[0], parts[1:]...)
	cmd.Dir = cwd
	cmd.Env = append(os.Environ(), "PYTHONUTF8=1", "PYTHONIOENCODING=utf-8")

	stdinPipe, err := cmd.StdinPipe()
	if err != nil {
		return fmt.Errorf("建立 stdin pipe 失敗: %w", err)
	}
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("建立 stdout pipe 失敗: %w", err)
	}
	cmd.Stderr = cmd.Stdout // merge stderr into stdout

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("啟動程序失敗: %w", err)
	}

	mp.cmd = cmd
	mp.stdin = stdinPipe
	mp.stdout = stdoutPipe
	mp.outputLines = make([]string, 0, 256)
	mp.pendingResp = make([]string, 0, 256)
	mp.responseCh = make(chan string, 1)

	go mp.readOutput()

	log.Printf("[process] %s 已啟動 (pid=%d)", mp.Name, cmd.Process.Pid)
	return nil
}

// readOutput reads stdout in chunks (not line-by-line) to handle \r TUI updates.
func (mp *ManagedProcess) readOutput() {
	reader := bufio.NewReaderSize(mp.stdout, 8192)
	buf := make([]byte, 4096)

	var remainder string

	for {
		n, err := reader.Read(buf)
		if n > 0 {
			text := remainder + string(buf[:n])
			remainder = ""

			// Split on newlines and carriage returns
			for len(text) > 0 {
				nlIdx := strings.IndexByte(text, '\n')
				crIdx := strings.IndexByte(text, '\r')

				var pos int
				if nlIdx == -1 && crIdx == -1 {
					// No line break found, save as remainder
					remainder = text
					break
				} else if nlIdx == -1 {
					pos = crIdx
				} else if crIdx == -1 {
					pos = nlIdx
				} else {
					if nlIdx < crIdx {
						pos = nlIdx
					} else {
						pos = crIdx
					}
				}

				line := strings.TrimRight(text[:pos], " \t")
				text = text[pos+1:]

				if line == "" {
					continue
				}

				clean := StripANSI(line)
				if clean == "" {
					continue
				}

				mp.mu.Lock()
				mp.outputLines = append(mp.outputLines, clean)
				if len(mp.outputLines) > maxCaptureLines {
					mp.outputLines = mp.outputLines[len(mp.outputLines)-maxCaptureLines:]
				}
				if mp.collecting {
					mp.pendingResp = append(mp.pendingResp, clean)
					// Check for response completion marker
					if strings.Contains(clean, "▸ Time:") || strings.Contains(clean, "Time:") {
						response := mp.extractFinalResponse(mp.pendingResp)
						mp.pendingResp = mp.pendingResp[:0]
						mp.collecting = false
						mp.mu.Unlock()
						// Non-blocking send
						select {
						case mp.responseCh <- response:
						default:
						}
						continue
					}
				}
				mp.mu.Unlock()
			}
		}
		if err != nil {
			// Flush remainder
			if remainder != "" {
				clean := StripANSI(strings.TrimSpace(remainder))
				if clean != "" {
					mp.mu.Lock()
					mp.outputLines = append(mp.outputLines, clean)
					if mp.collecting {
						mp.pendingResp = append(mp.pendingResp, clean)
					}
					mp.mu.Unlock()
				}
			}
			break
		}
	}
}

// SendInput writes text to the subprocess stdin.
func (mp *ManagedProcess) SendInput(text string) error {
	if mp.stdin == nil {
		return fmt.Errorf("程序 %s 沒有 stdin", mp.Name)
	}

	mp.mu.Lock()
	mp.pendingResp = mp.pendingResp[:0]
	mp.collecting = true
	// Drain any old response
	select {
	case <-mp.responseCh:
	default:
	}
	mp.mu.Unlock()

	_, err := fmt.Fprintf(mp.stdin, "%s\n", text)
	return err
}

// WaitResponse waits for the response to complete (sees "▸ Time:" marker).
// Times out after 15 minutes.
func (mp *ManagedProcess) WaitResponse() string {
	select {
	case resp := <-mp.responseCh:
		return resp
	case <-time.After(responseTimeout):
		mp.mu.Lock()
		lines := make([]string, len(mp.pendingResp))
		copy(lines, mp.pendingResp)
		mp.pendingResp = mp.pendingResp[:0]
		mp.collecting = false
		mp.mu.Unlock()

		if len(lines) > 0 {
			return mp.extractFinalResponse(lines)
		}
		return timeoutMessage
	}
}

// Capture returns the last N lines of output.
func (mp *ManagedProcess) Capture(lines int) string {
	mp.mu.Lock()
	defer mp.mu.Unlock()

	if lines <= 0 || lines > len(mp.outputLines) {
		lines = len(mp.outputLines)
	}
	start := len(mp.outputLines) - lines
	if start < 0 {
		start = 0
	}
	return strings.Join(mp.outputLines[start:], "\n")
}

// IsAlive checks if the subprocess is still running.
func (mp *ManagedProcess) IsAlive() bool {
	if mp.cmd == nil || mp.cmd.Process == nil {
		return false
	}
	// On Unix, check if process exists; on Windows, ProcessState is nil if running
	if mp.cmd.ProcessState != nil {
		return false
	}
	return true
}

// PID returns the process ID.
func (mp *ManagedProcess) PID() int {
	if mp.cmd != nil && mp.cmd.Process != nil {
		return mp.cmd.Process.Pid
	}
	return 0
}

// Kill terminates the subprocess.
func (mp *ManagedProcess) Kill() {
	if mp.stdin != nil {
		mp.stdin.Close()
	}
	if mp.cmd != nil && mp.cmd.Process != nil {
		mp.cmd.Process.Kill()
		mp.cmd.Wait()
	}
}

// StripANSI removes ANSI escape sequences, terminal control codes, and spinner animations.
func StripANSI(text string) string {
	text = ansiRegex.ReplaceAllString(text, "")
	text = ansiQRegex.ReplaceAllString(text, "")
	text = spinnerRegex.ReplaceAllString(text, "")
	text = brailleRegex.ReplaceAllString(text, "")
	text = multiSpace.ReplaceAllString(text, " ")
	return strings.TrimSpace(text)
}

// extractFinalResponse filters kiro-cli output noise and returns the final answer.
func (mp *ManagedProcess) extractFinalResponse(responseLines []string) string {
	var responseBlocks []string
	var currentBlock []string
	inResponse := false

	for _, line := range responseLines {
		// Skip UI noise
		if containsAny(line, skipPatterns) {
			if inResponse && len(currentBlock) > 0 {
				responseBlocks = append(responseBlocks, strings.Join(currentBlock, "\n"))
				currentBlock = currentBlock[:0]
				inResponse = false
			}
			continue
		}

		// Skip JSON-like content (tool parameters)
		stripped := strings.TrimSpace(line)
		if stripped == "{" || stripped == "}" || stripped == "[" || stripped == "]" {
			continue
		}
		if strings.HasPrefix(stripped, "\"") && strings.HasSuffix(stripped, "\"") {
			continue
		}
		if strings.HasPrefix(stripped, "\"") && strings.Contains(stripped, "\":") {
			continue
		}

		// Lines starting with > are response content
		if strings.HasPrefix(line, "> ") {
			inResponse = true
			currentBlock = append(currentBlock, line[2:])
		} else if strings.HasPrefix(line, ">") {
			inResponse = true
			currentBlock = append(currentBlock, strings.TrimLeft(line[1:], " "))
		} else if inResponse && strings.TrimSpace(line) != "" {
			currentBlock = append(currentBlock, line)
		} else if strings.TrimSpace(line) == "" {
			if inResponse && len(currentBlock) > 0 {
				currentBlock = append(currentBlock, "")
			}
		}
	}

	if len(currentBlock) > 0 {
		responseBlocks = append(responseBlocks, strings.Join(currentBlock, "\n"))
	}

	// Return the longest block as the final answer
	if len(responseBlocks) > 0 {
		best := responseBlocks[0]
		for _, block := range responseBlocks[1:] {
			if len(block) > len(best) {
				best = block
			}
		}
		return strings.TrimSpace(best)
	}

	// Fallback: return all non-noise lines
	var filtered []string
	for _, line := range responseLines {
		if containsAny(line, skipPatterns) {
			continue
		}
		stripped := strings.TrimSpace(line)
		if stripped == "" || stripped == "{" || stripped == "}" || stripped == "[" || stripped == "]" {
			continue
		}
		if strings.HasPrefix(stripped, "\"") && (strings.Contains(stripped, "\":") || strings.HasSuffix(stripped, "\"")) {
			continue
		}
		filtered = append(filtered, line)
	}

	return strings.Join(filtered, "\n")
}

// containsAny checks if text contains any of the patterns.
func containsAny(text string, patterns []string) bool {
	for _, p := range patterns {
		if strings.Contains(text, p) {
			return true
		}
	}
	return false
}

// parseCommand splits a command string into executable and arguments,
// handling quoted paths (e.g., "C:\path with spaces\kiro-cli.exe" chat --flag).
func parseCommand(command string) []string {
	var parts []string
	var current strings.Builder
	inQuote := false

	for _, r := range command {
		switch {
		case r == '"':
			inQuote = !inQuote
		case r == ' ' && !inQuote:
			if current.Len() > 0 {
				parts = append(parts, current.String())
				current.Reset()
			}
		default:
			current.WriteRune(r)
		}
	}
	if current.Len() > 0 {
		parts = append(parts, current.String())
	}
	return parts
}

File: src/kiro_multi_agent/process.py
"""Process management for Kiro CLI instances (asyncio subprocess, Windows-native)."""
from __future__ import annotations

import asyncio
import logging
import os
import sys
from dataclasses import dataclass, field

log = logging.getLogger(__name__)


@dataclass
class ManagedProcess:
    """Wraps an asyncio subprocess with stdin/stdout access and output capture."""

    name: str
    proc: asyncio.subprocess.Process | None = None
    _output_lines: list[str] = field(default_factory=list)
    _reader_task: asyncio.Task | None = None
    _max_capture: int = 5000
    _on_output: object = None  # Callable[[str, str], Awaitable[None]] | None
    _pending_response: list[str] = field(default_factory=list)
    _response_event: asyncio.Event | None = None
    _collecting_response: bool = False

    @property
    def pid(self) -> int | None:
        return self.proc.pid if self.proc else None

    def is_alive(self) -> bool:
        return self.proc is not None and self.proc.returncode is None

    async def start(self, cmd: str, cwd: str) -> None:
        full_env = {**os.environ, "PYTHONUTF8": "1", "PYTHONIOENCODING": "utf-8"}
        kwargs = dict(
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            cwd=cwd,
            env=full_env,
        )
        if sys.platform == "win32":
            kwargs["creationflags"] = 0x00000200  # CREATE_NEW_PROCESS_GROUP
        self.proc = await asyncio.create_subprocess_shell(cmd, **kwargs)
        self._output_lines.clear()
        self._reader_task = asyncio.create_task(self._read_output())
        log.info("Process started: %s (pid=%s)", self.name, self.pid)

    async def _read_output(self) -> None:
        assert self.proc and self.proc.stdout
        try:
            buffer = ""
            while True:
                chunk = await self.proc.stdout.read(4096)
                if not chunk:
                    break
                buffer += chunk.decode("utf-8", errors="replace")
                # Split on newlines and carriage returns
                while "\n" in buffer or "\r" in buffer:
                    # Find earliest line break
                    nl = buffer.find("\n")
                    cr = buffer.find("\r")
                    if nl == -1:
                        pos = cr
                    elif cr == -1:
                        pos = nl
                    else:
                        pos = min(nl, cr)
                    line = buffer[:pos].rstrip()
                    buffer = buffer[pos + 1:]
                    # Skip empty lines from consecutive \r\n
                    if line:
                        # Strip ANSI escape codes for clean text
                        clean = self._strip_ansi(line)
                        if clean:
                            self._output_lines.append(clean)
                            if len(self._output_lines) > self._max_capture:
                                self._output_lines = self._output_lines[-self._max_capture:]
                            # Notify output callback if set
                            if self._on_output and self._collecting_response:
                                self._pending_response.append(clean)
            # Flush remaining buffer
            if buffer.strip():
                clean = self._strip_ansi(buffer.strip())
                if clean:
                    self._output_lines.append(clean)
                    if self._on_output and self._collecting_response:
                        self._pending_response.append(clean)
        except (asyncio.CancelledError, OSError):
            pass

    @staticmethod
    def _strip_ansi(text: str) -> str:
        """Remove ANSI escape sequences, terminal control codes, and spinner animations."""
        import re
        # Remove ANSI escape sequences (colors, cursor movement, etc.)
        text = re.sub(r"\x1b\[[0-9;]*[a-zA-Z]", "", text)
        # Remove ANSI escape sequences with ? prefix (cursor show/hide, etc.)
        text = re.sub(r"\x1b\[\?[0-9;]*[a-zA-Z]", "", text)
        # Remove spinner patterns (braille dots + "Thinking...")
        text = re.sub(r"[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]\s*Thinking\.\.\.", "", text)
        # Remove other spinner patterns
        text = re.sub(r"[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]\s*\w+\.\.\.", "", text)
        # Remove remaining braille spinner chars at start of line
        text = re.sub(r"^[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]+\s*", "", text)
        # Clean up multiple spaces
        text = re.sub(r"  +", " ", text)
        return text.strip()

    async def send_input(self, text: str) -> None:
        if not self.proc or not self.proc.stdin:
            raise RuntimeError(f"Process {self.name} has no stdin")
        # Start collecting response
        self._pending_response.clear()
        self._collecting_response = True
        self.proc.stdin.write((text + "\n").encode("utf-8"))
        await self.proc.stdin.drain()

    async def wait_response(self, timeout: float = 900) -> str:
        """Wait for the response to complete.

        Monitors output until kiro-cli prints the time marker (▸ Time:)
        which indicates the response is done. Times out after 15 minutes.
        """
        import logging
        import time
        _log = logging.getLogger(__name__)
        deadline = time.time() + timeout
        last_log = time.time()

        while time.time() < deadline:
            # Check if we see the end-of-response marker
            for line in self._pending_response:
                if "▸ Time:" in line or "Time:" in line:
                    # Response is complete
                    self._collecting_response = False
                    response_lines = self._pending_response.copy()
                    self._pending_response.clear()
                    return self._extract_final_response(response_lines)
            # Log progress every 30 seconds
            if time.time() - last_log > 30:
                _log.info(
                    "[wait] %s: %d lines buffered, last: %s",
                    self.name,
                    len(self._pending_response),
                    self._pending_response[-1][:80] if self._pending_response else "(empty)",
                )
                last_log = time.time()
            await asyncio.sleep(1)

        # Timeout - return whatever we have
        self._collecting_response = False
        response_lines = self._pending_response.copy()
        self._pending_response.clear()

        if response_lines:
            return self._extract_final_response(response_lines)
        return "⏱️ 搜尋超時（超過 15 分鐘），請稍後再試或換個問法。"

    def _extract_final_response(self, response_lines: list[str]) -> str:
        """Extract the final response from kiro-cli output, filtering noise."""

        # Filter out kiro-cli UI noise and extract only the final response
        filtered = []
        skip_patterns = [
            "▸ Time:", "╭─", "╰─", "│", "Model:",
            "Did you know?", "/context", "/model",
            "All tools are now trusted",
            "ask a question",
            "Thinking...",
            "Working...",
            "⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏",
            "[?25", "[?12",
            "Running tool", "using tool:",
            "Completed in", "- Completed",
            "Successfully read", "Reading file:",
            "Reading directory:", "Querying available",
            "I will run the following command:",
            "Purpose:", "max depth:", "max entries:",
            "(from mcp server:",
            "⋮", "with the param",
        ]

        # Find the last ">" prefixed response block
        response_blocks = []
        current_block = []
        in_response = False

        for line in response_lines:
            # Skip UI noise
            if any(skip in line for skip in skip_patterns):
                if in_response and current_block:
                    response_blocks.append("\n".join(current_block))
                    current_block = []
                    in_response = False
                continue

            # Skip JSON-like content (tool parameters)
            stripped = line.strip()
            if stripped in ("{", "}", "[", "]") or stripped.startswith('"') and stripped.endswith('"'):
                continue
            if stripped.startswith('"') and '":' in stripped:
                continue

            # Lines starting with > or following a > line are response content
            if line.startswith("> "):
                in_response = True
                current_block.append(line[2:])  # Remove "> " prefix
            elif line.startswith(">"):
                in_response = True
                current_block.append(line[1:].lstrip())
            elif in_response and line.strip():
                current_block.append(line)
            elif not line.strip():
                if in_response and current_block:
                    current_block.append("")  # Preserve blank lines in response

        if current_block:
            response_blocks.append("\n".join(current_block))

        # Return the last response block (the final answer)
        if response_blocks:
            # Pick the longest block as the final answer
            best = max(response_blocks, key=len)
            return best.strip()

        # Fallback: return all non-noise lines
        for line in response_lines:
            if not any(skip in line for skip in skip_patterns) and line.strip():
                stripped = line.strip()
                if stripped in ("{", "}", "[", "]"):
                    continue
                if stripped.startswith('"') and ('":' in stripped or stripped.endswith('"')):
                    continue
                filtered.append(line)

        return "\n".join(filtered)

    def capture(self, lines: int = 200) -> str:
        return "\n".join(self._output_lines[-lines:])

    async def kill(self) -> None:
        if self._reader_task:
            self._reader_task.cancel()
            self._reader_task = None
        if self.proc and self.proc.returncode is None:
            try:
                self.proc.terminate()
                await asyncio.wait_for(self.proc.wait(), timeout=5)
            except (asyncio.TimeoutError, OSError):
                self.proc.kill()
        self.proc = None


# Registry
_active: dict[str, ManagedProcess] = {}

def register(mp: ManagedProcess) -> None:
    _active[mp.name] = mp

def unregister(name: str) -> None:
    _active.pop(name, None)

def list_active() -> list[str]:
    return [name for name, mp in _active.items() if mp.is_alive()]
